### DISCUSSION TITLE
How does an upgrade qos work?

### DISCUSSION THREAD

**Discussion by YraganTron** [emqx/emqx#12213] [Q&A] [answered]
What happens if the producer sends a message via qos 0 and it is sent twice to the broker.
And the subscriber subscribes to a topic with qos 2 and the upgrade qos setting enabled. Will he receive the message twice or once?

**Comment by thalesmg** [comment, marked as answer]
Hi.

If the publishing client does publish two QoS 0 messages, they are considered independent, and will result in two messages being received by the subscriber.  Note that QoS 0 `PUBLISH` packets do not possess packet ids [[MQTT-2.2.1-2]](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#_Toc3901026).

Only if the publisher were to attempt publishing the same message with QoS 2 they would be deduped.  This correlation between the two attempts would be represented by the publisher by using the same packet id of the message for which a `PUBREC` has been receive.

This may be useful: 
- https://www.emqx.com/en/blog/introduction-to-mqtt-qos
